packages/experience-builder-sdk/src/blocks/editor/VisualEditorContext.tsx

```diff
@@ -86,7 +86,10 @@ export function VisualEditorContextProvider({
   const [isDragging, setIsDragging] = useState(false);
   const selectedNodeId = useRef<string>('');
   const [locale, setLocale] = useState<string>(initialLocale);
-  const [areInitialEntitiesFetched, setInitialEntitiesFetched] = useState(false);
+  // Set to true when entities were fetched from the parent app.
+  // Reset to false when we receive a tree update and need to validate
+  // again whether all necessary entities are fetched.
+  const [areEntitiesFetched, setEntitiesFetched] = useState(false);
   const [isFetchingEntities, setFetchingEntities] = useState(false);
```

Comment:
Maybe instead of two states we can use a status state that hold enums like `FETCHING` `ISFETCHED` etc?

Reply:
Good question. I was considering this as well. The reason why I didn't go for this was that it wouldn't have the same behavior or less readability IMO. Here are four different cases possible and all of them will happen throughout the lifetime of the component.
1. `areEntitiesFetched=false` and `isFetchingEntities=false` -> initial state; we can start fetching in the effect; don't render bindings or design components yet
2. `areEntitiesFetched=false` and `isFetchingEntities=true` -> we're fetching, so don't trigger another fetch call
3. `areEntitiesFetched=true` and `isFetchingEntities=false` -> render bindings & design components; when necessary trigger another fetch call
4. `areEntitiesFetched=true` and `isFetchingEntities=true` -> still allow rendering while refetching. This can happen when bindings or design components change.

So the enum would have four states. Giving them descriptive names might be a bit harder and it's probably not as readable as these explicit booleans. Note that there is no performance issue in having separate state variables and setting them one by one. Setting both booleans sequentially will still end in only one re-rendering.

Reply:
Ahh, I thought they would be exclusive so `isFetchingEntities` can never be true when `areEntitiesFetched ` is also true.